File: backend/src/controllers/userController.ts
import { Request, Response } from 'express';
import { prisma } from '../lib/db';
import { AuthenticatedRequest } from '../middleware/auth';
import { onlineUsersService } from '../realtime/onlineUsers';

export const getUserProfile = async (req: Request, res: Response) => {
  try {
    const userId = (req as any).user?.id;

    const user = await prisma.user.findUnique({
      where: { id: userId },
      select: {
        id: true,
        username: true,
        email: true,
        discriminator: true,
        avatar: true,
        status: true,
        bio: true,
        location: true,
        createdAt: true,
        password: false,
      }
    });

    if (!user) {
      return res.status(404).json({ error: 'User not found' });
    }

    return res.json({ user });

  } catch (error) {
    console.error('Get user profile error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
};

export const updateProfile = async (req: Request, res: Response) => {
  try {
    const userId = (req as any).user?.id;
    const { username, bio, location, avatar } = req.body;

    // Check if username is already taken by another user
    if (username) {
      const existingUser = await prisma.user.findFirst({
        where: {
          username,
          NOT: { id: userId }
        }
      });

      if (existingUser) {
        return res.status(400).json({
          error: 'Username already taken'
        });
      }
    }

    const updatedUser = await prisma.user.update({
      where: { id: userId },
      data: {
        ...(username && { username }),
        ...(bio !== undefined && { bio }),
        ...(location !== undefined && { location }),
        ...(avatar && { avatar }),
        updatedAt: new Date(),
      },
      select: {
        id: true,
        username: true,
        email: true,
        discriminator: true,
        avatar: true,
        status: true,
        bio: true,
        location: true,
        createdAt: true,
        password: false,
      }
    });

    return res.json({
      message: 'Profile updated successfully',
      user: updatedUser
    });

  } catch (error) {
    console.error('Update profile error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
};

export const getUsers = async (req: Request, res: Response) => {
  try {
    const { search, limit = 20, offset = 0 } = req.query;
    const currentUserId = (req as any).user?.id;

    const whereClause: any = {
      NOT: { id: currentUserId } // Exclude current user
    };

    if (search) {
      whereClause.OR = [
        { username: { contains: search as string, mode: 'insensitive' } },
        { email: { contains: search as string, mode: 'insensitive' } }
      ];
    }

    const users = await prisma.user.findMany({
      where: whereClause,
      select: {
        id: true,
        username: true,
        discriminator: true,
        avatar: true,
        status: true,
        bio: true,
        createdAt: true,
        password: false,
      },
      take: Number(limit),
      skip: Number(offset),
      orderBy: { createdAt: 'desc' }
    });

    const total = await prisma.user.count({ where: whereClause });

    res.json({
      users,
      pagination: {
        total,
        limit: Number(limit),
        offset: Number(offset),
        hasMore: Number(offset) + Number(limit) < total
      }
    });

  } catch (error) {
    console.error('Get users error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
};

/**
 * Get all online users
 */
export const getOnlineUsers = async (req: AuthenticatedRequest, res: Response) => {
  try {
    const onlineUsers = onlineUsersService.getOnlineUsers();
    
    res.json({
      onlineUsers,
      count: onlineUsers.length,
      totalOnline: onlineUsersService.getOnlineUsersCount()
    });
  } catch (error) {
    console.error('Get online users error:', error);
    res.status(500).json({ error: 'Failed to get online users' });
  }
};

/**
 * Get online users for a specific server
 */
export const getOnlineUsersForServer = async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { serverId } = req.params;
    const userId = req.user?.id;

    if (!userId) {
      return res.status(401).json({ error: 'Unauthorized' });
    }

    // Check if user is a member of the server
    const serverMember = await prisma.serverMember.findUnique({
      where: {
        serverId_userId: {
          serverId,
          userId
        }
      }
    });

    if (!serverMember) {
      return res.status(403).json({ error: 'Not a member of this server' });
    }

    const onlineUsers = await onlineUsersService.getOnlineUsersForServer(serverId);
    
    return res.json({
      onlineUsers,
      count: onlineUsers.length
    });
  } catch (error) {
    console.error('Get online users for server error:', error);
    return res.status(500).json({ error: 'Failed to get online users for server' });
  }
};

/**
 * Check if a specific user is online
 */
export const isUserOnline = async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { userId } = req.params;
    
    const isOnline = onlineUsersService.isUserOnline(userId);
    
    res.json({
      userId,
      isOnline
    });
  } catch (error) {
    console.error('Check user online status error:', error);
    res.status(500).json({ error: 'Failed to check user online status' });
  }
}; 

File: backend-go/internal/controllers/user_controller.go
package controllers

import (
	"net/http"

	"discord-clone-backend/internal/db"
	"discord-clone-backend/internal/middleware"
	"discord-clone-backend/pkg/database"

	"github.com/gin-gonic/gin"
)

// UserController handles user-related requests
type UserController struct {
	queries *db.Queries
}

// NewUserController creates a new user controller
func NewUserController() *UserController {
	return &UserController{
		queries: db.New(database.GetDB()),
	}
}

// GetUsers returns a list of users
func (uc *UserController) GetUsers(c *gin.Context) {
	ctx := c.Request.Context()

	users, err := uc.queries.GetUsers(ctx)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "Failed to fetch users",
		})
		return
	}

	// Convert to response format (excluding password)
	usersResp := make([]gin.H, len(users))
	for i, user := range users {
		usersResp[i] = gin.H{
			"id":            user.ID,
			"username":      user.Username,
			"email":         user.Email,
			"discriminator": user.Discriminator,
			"avatar":        database.PgTextToString(user.Avatar),
			"status":        database.PgTextToString(user.Status),
			"bio":           database.PgTextToString(user.Bio),
			"location":      database.PgTextToString(user.Location),
			"created_at":    user.CreatedAt,
			"updated_at":    user.UpdatedAt,
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"users": usersResp,
	})
}

// GetUser returns a specific user by ID
func (uc *UserController) GetUser(c *gin.Context) {
	ctx := c.Request.Context()
	userID := c.Param("id")

	user, err := uc.queries.GetUserByID(ctx, userID)
	if err != nil {
		if database.IsNoRowsError(err) {
			c.JSON(http.StatusNotFound, gin.H{
				"error": "User not found",
			})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "Failed to fetch user",
		})
		return
	}

	// Return user data (excluding password)
	userResp := gin.H{
		"id":            user.ID,
		"username":      user.Username,
		"email":         user.Email,
		"discriminator": user.Discriminator,
		"avatar":        database.PgTextToString(user.Avatar),
		"status":        database.PgTextToString(user.Status),
		"bio":           database.PgTextToString(user.Bio),
		"location":      database.PgTextToString(user.Location),
		"created_at":    user.CreatedAt,
		"updated_at":    user.UpdatedAt,
	}

	c.JSON(http.StatusOK, gin.H{
		"user": userResp,
	})
}

// UpdateUser updates a user's information
func (uc *UserController) UpdateUser(c *gin.Context) {
	ctx := c.Request.Context()
	userID := c.Param("id")
	currentUserID, _ := middleware.GetUserID(c)

	// Users can only update their own profile
	if userID != currentUserID {
		c.JSON(http.StatusForbidden, gin.H{
			"error": "You can only update your own profile",
		})
		return
	}

	var updateData struct {
		Username *string `json:"username"`
		Bio      *string `json:"bio"`
		Location *string `json:"location"`
		Avatar   *string `json:"avatar"`
	}

	if err := c.ShouldBindJSON(&updateData); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "Invalid request data",
		})
		return
	}

	// Get current user to preserve fields that aren't being updated
	currentUser, err := uc.queries.GetUserByID(ctx, userID)
	if err != nil {
		if database.IsNoRowsError(err) {
			c.JSON(http.StatusNotFound, gin.H{
				"error": "User not found",
			})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "Failed to fetch user",
		})
		return
	}

	// Update only provided fields, preserve others
	username := currentUser.Username
	if updateData.Username != nil {
		username = *updateData.Username
	}

	email := currentUser.Email // Email shouldn't be updated via this endpoint, but preserve it

	avatar := database.PgTextToString(currentUser.Avatar)
	if updateData.Avatar != nil {
		avatar = updateData.Avatar
	}

	status := database.PgTextToString(currentUser.Status)
	if status == nil {
		statusStr := "offline"
		status = &statusStr
	}

	bio := database.PgTextToString(currentUser.Bio)
	if updateData.Bio != nil {
		bio = updateData.Bio
	}

	location := database.PgTextToString(currentUser.Location)
	if updateData.Location != nil {
		location = updateData.Location
	}

	// Update user
	updatedUser, err := uc.queries.UpdateUser(ctx, db.UpdateUserParams{
		ID:       userID,
		Username: username,
		Email:    email,
		Avatar:   database.StringToPgText(avatar),
		Status:   database.StringToPgText(status),
		Bio:      database.StringToPgText(bio),
		Location: database.StringToPgText(location),
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "Failed to update user",
		})
		return
	}

	// Return updated user data (excluding password)
	userResp := gin.H{
		"id":            updatedUser.ID,
		"username":      updatedUser.Username,
		"email":         updatedUser.Email,
		"discriminator": updatedUser.Discriminator,
		"avatar":        database.PgTextToString(updatedUser.Avatar),
		"status":        database.PgTextToString(updatedUser.Status),
		"bio":           database.PgTextToString(updatedUser.Bio),
		"location":      database.PgTextToString(updatedUser.Location),
		"created_at":    updatedUser.CreatedAt,
		"updated_at":    updatedUser.UpdatedAt,
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "User updated successfully",
		"user":    userResp,
	})
}


// GetOnlineUsers returns all online users
func (uc *UserController) GetOnlineUsers(c *gin.Context) {
	ctx := c.Request.Context()

	// Get all users
	users, err := uc.queries.GetUsers(ctx)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "Failed to fetch users",
		})
		return
	}

	// Filter online users and convert to response format
	onlineUsers := []gin.H{}
	for _, user := range users {
		status := database.PgTextToString(user.Status)
		if status != nil && *status == "online" {
			onlineUsers = append(onlineUsers, gin.H{
				"userId":       user.ID,
				"username":     user.Username,
				"discriminator": user.Discriminator,
				"avatar":       database.PgTextToString(user.Avatar),
			})
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"onlineUsers": onlineUsers,
		"count":       len(onlineUsers),
		"totalOnline": len(onlineUsers),
	})
}

// GetOnlineUsersForServer returns online users for a specific server
func (uc *UserController) GetOnlineUsersForServer(c *gin.Context) {
	ctx := c.Request.Context()
	serverID := c.Param("serverId")
	userID, _ := middleware.GetUserID(c)

	// Check if user is a member of the server
	_, err := uc.queries.GetServerMember(ctx, db.GetServerMemberParams{
		ServerID: serverID,
		UserID:   userID,
	})
	if err != nil {
		if database.IsNoRowsError(err) {
			c.JSON(http.StatusForbidden, gin.H{
				"error": "Not a member of this server",
			})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "Failed to check membership",
		})
		return
	}

	// Get server members with user info
	serverMembers, err := uc.queries.GetServerMembersWithUsers(ctx, serverID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "Failed to fetch server members",
		})
		return
	}

	// Filter online users
	onlineUsers := []gin.H{}
	for _, member := range serverMembers {
		// Check if user status is online
		// Note: The GetServerMembersWithUsers query returns user fields, but we need to check status
		// Since the query doesn't return status, we'll get the user separately or modify the query
		// For now, let's get the user to check status
		user, err := uc.queries.GetUserByID(ctx, member.UserID)
		if err != nil {
			continue // Skip if user not found
		}

		status := database.PgTextToString(user.Status)
		if status != nil && *status == "online" {
			onlineUsers = append(onlineUsers, gin.H{
				"userId":       user.ID,
				"username":     user.Username,
				"discriminator": user.Discriminator,
				"avatar":       database.PgTextToString(user.Avatar),
			})
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"onlineUsers": onlineUsers,
		"count":       len(onlineUsers),
	})
}

// IsUserOnline checks if a specific user is online
func (uc *UserController) IsUserOnline(c *gin.Context) {
	ctx := c.Request.Context()
	userID := c.Param("userId")

	user, err := uc.queries.GetUserByID(ctx, userID)
	if err != nil {
		if database.IsNoRowsError(err) {
			c.JSON(http.StatusNotFound, gin.H{
				"error": "User not found",
			})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "Failed to fetch user",
		})
		return
	}

	status := database.PgTextToString(user.Status)
	isOnline := status != nil && *status == "online"

	c.JSON(http.StatusOK, gin.H{
		"userId":  userID,
		"isOnline": isOnline,
	})
}